File: typescript/server/src/agentConfig.ts
/**
 * SMOODEV-590 — Per-agent configuration (TypeScript server port).
 *
 * The server resolves each conversation's `agentId` into an {@link AgentConfig}
 * (the agent's own `instructions`, `conversationWorkflow`, optional greeting /
 * personality / tool allow-list), and folds it into the system prompt for that
 * agent's turns — so two agents in the same org behave differently instead of all
 * using one generic org persona.
 *
 * The delivery seam is an {@link AgentConfigResolver}, mirroring the server's other
 * pluggable seams ({@link AuthVerifier}, {@link AccessKnowledge}): the reference
 * ships an in-memory / no-op resolver; a real deployment plugs in one backed by the
 * monorepo `agents` table. The `create_conversation_session` payload carries only an
 * `agentId` (per the spec), so config is resolved server-side by that id, never from
 * the wire frame.
 */
import { parseWorkflow, renderWorkflowPromptSection, type ConversationWorkflow } from './workflow.js';

/**
 * The per-agent config that shapes an agent's conversations. Every field is
 * optional — an agent may set only `instructions`, only a `conversationWorkflow`,
 * or nothing (in which case the server falls back to its base/org prompt).
 */
export interface AgentConfig {
    /** Freeform system-prompt body for this agent (`agents.instructions.prompt`). */
    instructions?: string;
    /** Structured guided-agency workflow (`agents.conversation_workflow`). */
    conversationWorkflow?: ConversationWorkflow;
    /** Optional greeting to weave into the agent's first reply. */
    greeting?: string;
    /** Optional short personality descriptor folded into the persona section. */
    personality?: string;
    /** Optional allow-list of tool names this agent may use; when set, the server's
     *  tool set is filtered to it. Empty / undefined → all server tools available. */
    allowedTools?: string[];
}

/**
 * Resolves an `agentId` into its {@link AgentConfig}. `undefined` (agent unknown /
 * no per-agent config) → the server uses its base/org default prompt + tools, so
 * behavior is unchanged for un-configured agents.
 */
export interface AgentConfigResolver {
    resolve(agentId: string): Promise<AgentConfig | undefined> | AgentConfig | undefined;
}

/** A resolver backed by a fixed in-memory `agentId → AgentConfig` map. The reference
 *  implementation for tests / local use; a real deployment reads the `agents` table. */
export class StaticAgentConfigResolver implements AgentConfigResolver {
    private readonly byId: Map<string, AgentConfig>;

    constructor(configs: Record<string, AgentConfig> = {}) {
        this.byId = new Map(Object.entries(configs));
    }

    resolve(agentId: string): AgentConfig | undefined {
        return this.byId.get(agentId);
    }
}

/**
 * Tolerantly parse a raw agent record (the shape stored in the monorepo `agents`
 * table: `instructions` jsonb `{prompt}`, `conversation_workflow` jsonb, etc.) into
 * an {@link AgentConfig}. Malformed sub-fields are dropped individually — a broken
 * `conversation_workflow` doesn't discard a valid `instructions.prompt` — and the
 * function never throws, so a bad record degrades gracefully. Returns `undefined`
 * only when nothing usable is present.
 */
export function parseAgentConfig(raw: unknown): AgentConfig | undefined {
    if (typeof raw !== 'object' || raw === null || Array.isArray(raw)) return undefined;
    const obj = raw as Record<string, unknown>;
    const config: AgentConfig = {};

    // instructions: either the jsonb `{ prompt: string }` or a bare string.
    const instr = obj.instructions;
    if (typeof instr === 'string' && instr.trim().length > 0) {
        config.instructions = instr;
    } else if (typeof instr === 'object' && instr !== null) {
        const prompt = (instr as Record<string, unknown>).prompt;
        if (typeof prompt === 'string' && prompt.trim().length > 0) config.instructions = prompt;
    }

    // conversation_workflow (snake) / conversationWorkflow (camel) — tolerant parse.
    const workflow = parseWorkflow(obj.conversation_workflow ?? obj.conversationWorkflow);
    if (workflow) config.conversationWorkflow = workflow;

    if (typeof obj.greeting === 'string' && obj.greeting.trim().length > 0) config.greeting = obj.greeting;
    if (typeof obj.personality === 'string' && obj.personality.trim().length > 0) config.personality = obj.personality;

    const tools = obj.tool_config ?? obj.allowedTools;
    if (Array.isArray(tools)) {
        const names = tools.filter((t): t is string => typeof t === 'string' && t.length > 0);
        if (names.length > 0) config.allowedTools = names;
    }

    return Object.keys(config).length > 0 ? config : undefined;
}

/**
 * Assemble the effective system prompt for a turn from the server's base prompt, the
 * per-agent config, and the conversation's current workflow step.
 *
 * When `config` is undefined or empty this returns `base` unchanged (behavior is
 * identical to before per-agent config existed). Otherwise the agent's own
 * `instructions` become the primary body, augmented by the base prompt's grounding
 * rules, plus optional personality / greeting sections and the rendered workflow
 * step.
 */
export function assembleSystemPrompt(base: string, config: AgentConfig | undefined, currentStepId: string | null | undefined): string {
    if (!config) return base;

    const sections: string[] = [];

    if (config.personality) sections.push(`<Personality>\n${config.personality}\n</Personality>`);

    // The agent's own instructions are the primary persona; the base prompt's
    // grounding / behavior rules follow so they always apply.
    if (config.instructions) {
        sections.push(`<AgentInstructions>\n${config.instructions}\n</AgentInstructions>`);
        sections.push(base);
    } else {
        sections.push(base);
    }

    if (config.greeting) {
        sections.push(
            `<GreetingAwareness>\nIf this is your first reply in the conversation, open with a natural, brief variant of: "${config.greeting}" — then address the user's message. Do not repeat it verbatim on later turns.\n</GreetingAwareness>`,
        );
    }

    const workflowSection = renderWorkflowPromptSection(config.conversationWorkflow, currentStepId);
    if (workflowSection) sections.push(workflowSection);

    return sections.join('\n\n');
}

File: dotnet/server/src/AgentConfig.cs
using System.Text.Json;
using System.Text.Json.Nodes;

namespace SmooAI.SmoothOperator.Server;

/// <summary>
/// A single step in a structured conversation workflow — the C# analog of the monorepo's
/// <c>ConversationWorkflowStep</c> (packages/schemas/src/agents/agent.ts). The judge evaluates
/// <see cref="Criteria"/> after each turn to decide whether to advance to <see cref="Next"/> (or,
/// absent that, the next step in order).
/// </summary>
public sealed record ConversationWorkflowStep(string Id, string Intent, string Criteria, string? Next);

/// <summary>
/// A structured conversation workflow: an overall <see cref="Goal"/> plus ordered
/// <see cref="Steps"/>. When an agent has one, the current step's intent + criteria are rendered
/// into the system prompt and a post-turn judge advances the pointer. The C# analog of the
/// monorepo's <c>ConversationWorkflow</c>.
/// </summary>
public sealed record ConversationWorkflow(string Goal, IReadOnlyList<ConversationWorkflowStep> Steps);

/// <summary>
/// One entry of the agent's <c>tool_config.enabledTools</c> (mirrors the monorepo
/// <c>AgentToolConfig</c>). <see cref="ToolId"/> is snake_case (matches the runtime tool registry
/// keys). <see cref="AuthLevel"/> / <see cref="Config"/> are preserved for the host even though the
/// reference server doesn't act on them yet.
/// </summary>
public sealed record EnabledTool(string ToolId, bool Enabled, string AuthLevel, JsonObject? Config);

/// <summary>
/// Per-agent configuration resolved for a conversation (the analog of the monorepo <c>agents</c>
/// row's <c>instructions</c> / <c>conversation_workflow</c> / <c>greeting</c> / <c>personality</c>).
/// The server applies this ON TOP of the org/host default persona so each agent behaves as itself
/// rather than a generic customer-support persona.
///
/// Every field is optional: an agent with no config (or malformed config — see the tolerant
/// <c>Parse*</c> helpers) resolves to a fully-null instance, and the server falls back to its
/// existing default-persona behavior. Never throws on bad input — a broken jsonb blob degrades a
/// session to the default, it never crashes it.
/// </summary>
public sealed record AgentConfig(
    string? InstructionsPrompt = null,
    ConversationWorkflow? Workflow = null,
    string? Greeting = null,
    string? Personality = null,
    IReadOnlyList<EnabledTool>? EnabledTools = null,
    string? Visibility = null)
{
    /// <summary>An empty config — the "no per-agent overrides" sentinel.</summary>
    public static readonly AgentConfig Empty = new();

    /// <summary>True when this config carries nothing the server would apply. A non-null
    /// <see cref="EnabledTools"/> (even empty) is a tool restriction, so it counts as configured.</summary>
    public bool IsEmpty => string.IsNullOrWhiteSpace(InstructionsPrompt) && Workflow is null && string.IsNullOrWhiteSpace(Greeting) && string.IsNullOrWhiteSpace(Personality) && EnabledTools is null;

    /// <summary>
    /// Parse the <c>instructions</c> jsonb (<c>{"prompt": "..."}</c>) into the freeform prompt
    /// string. Tolerant: null/blank/malformed/absent-<c>prompt</c> → <c>null</c> (no override).
    /// A bare JSON string is also accepted (treated as the prompt) for forgiveness.
    /// </summary>
    public static string? ParseInstructions(string? json)
    {
        if (string.IsNullOrWhiteSpace(json))
        {
            return null;
        }
        try
        {
            var node = JsonNode.Parse(json);
            var prompt = node switch
            {
                JsonObject obj => obj["prompt"]?.GetValue<string>(),
                JsonValue val when val.TryGetValue<string>(out var s) => s,
                _ => null,
            };
            return string.IsNullOrWhiteSpace(prompt) ? null : prompt;
        }
        catch (Exception ex) when (ex is JsonException or FormatException or InvalidOperationException)
        {
            return null;
        }
    }

    /// <summary>
    /// Parse the <c>conversation_workflow</c> jsonb into a <see cref="ConversationWorkflow"/>.
    /// Tolerant of every malformed shape (null/blank/not-an-object/missing-goal/empty-steps/bad-step)
    /// → <c>null</c>, so a broken workflow silently degrades to freeform-prompt behavior rather than
    /// crashing the session. Enforces the schema bounds (1..20 steps; non-empty id/intent/criteria)
    /// so a partially-valid blob doesn't render a garbage prompt section.
    /// </summary>
    public static ConversationWorkflow? ParseWorkflow(string? json)
    {
        if (string.IsNullOrWhiteSpace(json))
        {
            return null;
        }
        try
        {
            if (JsonNode.Parse(json) is not JsonObject obj)
            {
                return null;
            }

            var goal = obj["goal"]?.GetValue<string>();
            if (string.IsNullOrWhiteSpace(goal))
            {
                return null;
            }

            if (obj["steps"] is not JsonArray stepsArray || stepsArray.Count is 0 or > 20)
            {
                return null;
            }

            var steps = new List<ConversationWorkflowStep>(stepsArray.Count);
            foreach (var element in stepsArray)
            {
                if (element is not JsonObject stepObj)
                {
                    return null;
                }
                var id = stepObj["id"]?.GetValue<string>();
                var intent = stepObj["intent"]?.GetValue<string>();
                var criteria = stepObj["criteria"]?.GetValue<string>();
                if (string.IsNullOrWhiteSpace(id) || string.IsNullOrWhiteSpace(intent) || string.IsNullOrWhiteSpace(criteria))
                {
                    return null;
                }
                var next = stepObj["next"]?.GetValue<string>();
                steps.Add(new ConversationWorkflowStep(id, intent, criteria, string.IsNullOrWhiteSpace(next) ? null : next));
            }

            return new ConversationWorkflow(goal, steps);
        }
        catch (Exception ex) when (ex is JsonException or FormatException or InvalidOperationException)
        {
            return null;
        }
    }

    /// <summary>
    /// Parse the agent's <c>tool_config</c> jsonb (<c>{ enabledTools: [{ toolId, enabled?, authLevel?,
    /// config? }] }</c>, per the monorepo <c>AgentToolConfig</c>) into the enabled-tool list used to
    /// restrict the server's tool set. Returns <c>null</c> = NO restriction (full tool set) when the
    /// config is missing/blank/malformed OR <c>enabledTools</c> is empty — empty must NOT mean
    /// "no tools" (every agent row defaults <c>enabledTools</c> to <c>[]</c>). A non-empty
    /// <c>enabledTools</c> yields a non-null list (even if all entries were dropped as malformed),
    /// so the restriction stays active and fails closed to least-privilege. Individual entries with a
    /// missing / non-snake_case <c>toolId</c> are dropped (tolerant); unknown toolIds are ignored at
    /// filter time by intersecting with the registered tools.
    /// </summary>
    public static IReadOnlyList<EnabledTool>? ParseToolConfig(string? json)
    {
        if (string.IsNullOrWhiteSpace(json))
        {
            return null;
        }
        try
        {
            if (JsonNode.Parse(json) is not JsonObject obj || obj["enabledTools"] is not JsonArray entries || entries.Count == 0)
            {
                return null;
            }

            var tools = new List<EnabledTool>(entries.Count);
            foreach (var element in entries)
            {
                if (element is not JsonObject entry)
                {
                    continue;
                }
                var toolId = entry["toolId"]?.GetValue<string>();
                // snake_case only — camelCase ids silently fail to bind at runtime (SMOODEV-981).
                if (string.IsNullOrWhiteSpace(toolId) || !IsSnakeCase(toolId))
                {
                    continue;
                }
                var enabled = entry["enabled"] is JsonValue e && e.TryGetValue<bool>(out var b) ? b : true;
                var authLevel = entry["authLevel"]?.GetValue<string>() ?? "none";
                tools.Add(new EnabledTool(toolId, enabled, string.IsNullOrWhiteSpace(authLevel) ? "none" : authLevel, entry["config"] as JsonObject));
            }
            // Non-empty enabledTools → keep the restriction active even if every entry was dropped.
            return tools;
        }
        catch (Exception ex) when (ex is JsonException or FormatException or InvalidOperationException)
        {
            return null;
        }
    }

    private static bool IsSnakeCase(string value)
    {
        foreach (var ch in value)
        {
            if (!(ch is >= 'a' and <= 'z' or >= '0' and <= '9' or '_'))
            {
                return false;
            }
        }
        return true;
    }
}
